File: impl/ts/src/store/blocklist.ts
/**
 * SQLite-backed BlockListLookup. Mirrors impl/go/internal/store/blocklist.go.
 *
 * @module
 */

import { randomBytes } from "node:crypto";

import type Database from "better-sqlite3";

import type {
  BlockEntry,
  BlockList,
  BlockListLookup,
  BlocklistEntity,
  BlocklistEntityType,
  BlocklistScope,
} from "@sempdev/semp/delivery";

export class SQLiteBlockList implements BlockListLookup {
  private readonly db: Database.Database;

  constructor(db: Database.Database) {
    this.db = db;
  }

  async lookup(recipient: string): Promise<BlockList | null> {
    const rows = this.db
      .prepare<
        [string],
        BlockRow
      >(
        `SELECT id, entity_type, entity_value, acknowledgment, reason, scope, created_at, expires_at
           FROM block_entries WHERE user_id = ?`,
      )
      .all(recipient);
    const entries: BlockEntry[] = rows.map((r) => rowToEntry(r));
    return {
      user_id: recipient,
      list_version: 1,
      entries,
    };
  }

  /** Insert a block entry. Returns its id (generated when omitted). */
  addEntry(userId: string, entry: BlockEntry): string {
    const id = entry.id !== "" ? entry.id : randomBytes(16).toString("hex");
    const now = isoNow();
    const expires = entry.expires_at !== undefined && entry.expires_at !== ""
      ? entry.expires_at
      : null;
    const entityValue = entityValueOf(entry.entity);
    this.db
      .prepare(
        `INSERT INTO block_entries (id, user_id, entity_type, entity_value, acknowledgment, reason, scope, created_at, expires_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      )
      .run(
        id,
        userId,
        entry.entity.type,
        entityValue,
        entry.acknowledgment,
        entry.reason ?? null,
        entry.scope,
        now,
        expires,
      );
    return id;
  }

  /** Remove a block entry by id. */
  removeEntry(entryId: string): void {
    this.db.prepare(`DELETE FROM block_entries WHERE id = ?`).run(entryId);
  }

  /** Snapshot the entries for a user. */
  async listEntries(userId: string): Promise<BlockEntry[]> {
    const list = await this.lookup(userId);
    return list?.entries ?? [];
  }
}

interface BlockRow {
  id: string;
  entity_type: string;
  entity_value: string;
  acknowledgment: string;
  reason: string | null;
  scope: string;
  created_at: string;
  expires_at: string | null;
}

function rowToEntry(r: BlockRow): BlockEntry {
  const entity: BlocklistEntity = { type: r.entity_type as BlocklistEntityType };
  switch (entity.type) {
    case "user":
      entity.address = r.entity_value;
      break;
    case "domain":
      entity.domain = r.entity_value;
      break;
    case "server":
      entity.hostname = r.entity_value;
      break;
  }
  const entry: BlockEntry = {
    id: r.id,
    entity,
    acknowledgment: (r.acknowledgment === "delivered" ||
    r.acknowledgment === "rejected" ||
    r.acknowledgment === "silent"
      ? r.acknowledgment
      : "rejected") as BlockEntry["acknowledgment"],
    scope: (r.scope === "all" || r.scope === "direct" || r.scope === "group"
      ? r.scope
      : "all") as BlocklistScope,
    created_at: r.created_at,
    created_by_device_id: "",
  };
  if (r.reason !== null) {
    entry.reason = r.reason;
  }
  if (r.expires_at !== null) {
    entry.expires_at = r.expires_at;
  }
  return entry;
}

function entityValueOf(entity: BlocklistEntity): string {
  switch (entity.type) {
    case "user":
      return entity.address ?? "";
    case "domain":
      return entity.domain ?? "";
    case "server":
      return entity.hostname ?? "";
  }
}

function isoNow(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

File: impl/go/internal/store/blocklist.go
package store

import (
	"context"
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"time"

	semp "semp.dev/semp-go"
	"semp.dev/semp-go/delivery"
)

// SQLiteBlockList implements delivery.BlockListLookup backed by SQLite.
type SQLiteBlockList struct {
	db *sql.DB
}

// NewSQLiteBlockList wraps an initialised database.
func NewSQLiteBlockList(db *sql.DB) *SQLiteBlockList {
	return &SQLiteBlockList{db: db}
}

// Lookup returns the block list for recipient.
func (bl *SQLiteBlockList) Lookup(ctx context.Context, recipient string) (*delivery.BlockList, error) {
	rows, err := bl.db.QueryContext(ctx,
		`SELECT id, entity_type, entity_value, acknowledgment, reason, scope, created_at, expires_at
		   FROM block_entries WHERE user_id = ?`, recipient)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	list := &delivery.BlockList{UserID: recipient}
	for rows.Next() {
		var (
			id, entType, entValue, ack, scope, createdStr string
			reason, expiresStr                            sql.NullString
		)
		if err := rows.Scan(&id, &entType, &entValue, &ack, &reason, &scope, &createdStr, &expiresStr); err != nil {
			return nil, err
		}
		created, _ := time.Parse(time.RFC3339, createdStr)
		entity := delivery.Entity{Type: delivery.EntityType(entType)}
		switch entity.Type {
		case delivery.EntityUser:
			entity.Address = entValue
		case delivery.EntityDomain:
			entity.Domain = entValue
		case delivery.EntityServer:
			entity.Hostname = entValue
		}
		entry := delivery.BlockEntry{
			ID:     id,
			Entity: entity,
			Acknowledgment: semp.Acknowledgment(ack),
			Scope:          delivery.Scope(scope),
			CreatedAt:      created,
		}
		if reason.Valid {
			entry.Reason = reason.String
		}
		if expiresStr.Valid {
			t, _ := time.Parse(time.RFC3339, expiresStr.String)
			entry.ExpiresAt = &t
		}
		list.Entries = append(list.Entries, entry)
	}
	return list, rows.Err()
}

// AddEntry adds a block entry for a user.
func (bl *SQLiteBlockList) AddEntry(ctx context.Context, userID string, entry delivery.BlockEntry) (string, error) {
	if entry.ID == "" {
		b := make([]byte, 16)
		_, _ = rand.Read(b)
		entry.ID = hex.EncodeToString(b)
	}
	now := time.Now().UTC().Format(time.RFC3339)
	var expires *string
	if entry.ExpiresAt != nil {
		s := entry.ExpiresAt.Format(time.RFC3339)
		expires = &s
	}
	var entityValue string
	switch entry.Entity.Type {
	case delivery.EntityUser:
		entityValue = entry.Entity.Address
	case delivery.EntityDomain:
		entityValue = entry.Entity.Domain
	case delivery.EntityServer:
		entityValue = entry.Entity.Hostname
	}
	_, err := bl.db.ExecContext(ctx,
		`INSERT INTO block_entries (id, user_id, entity_type, entity_value, acknowledgment, reason, scope, created_at, expires_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		entry.ID, userID, string(entry.Entity.Type), entityValue,
		string(entry.Acknowledgment), entry.Reason, string(entry.Scope), now, expires)
	return entry.ID, err
}

// RemoveEntry removes a block entry by ID.
func (bl *SQLiteBlockList) RemoveEntry(ctx context.Context, entryID string) error {
	_, err := bl.db.ExecContext(ctx, `DELETE FROM block_entries WHERE id = ?`, entryID)
	return err
}

// ListEntries returns all block entries for a user.
func (bl *SQLiteBlockList) ListEntries(ctx context.Context, userID string) ([]delivery.BlockEntry, error) {
	list, err := bl.Lookup(ctx, userID)
	if err != nil {
		return nil, err
	}
	return list.Entries, nil
}
